multicluster/controllers/multicluster/resourceexport_controller.go

```diff
@@ -47,16 +73,317 @@ type ResourceExportReconciler struct {
+	resImport *mcsv1alpha1.ResourceImport,
+	createResImport bool) (*mcsv1alpha1.ResourceImport, bool, error) {
+	newResImport := resImport.DeepCopy()
+	newResImport.Spec.Name = resExport.Spec.Name
+	newResImport.Spec.Namespace = resExport.Spec.Namespace
+	newResImport.Spec.Kind = common.EndpointsKind
+	if createResImport {
+		newResImport.Spec.Endpoints = &mcsv1alpha1.EndpointsImport{
+			Subsets: resExport.Spec.Endpoints.Subsets,
+		}
+		return newResImport, true, nil
+	}
+	// check all matched Endpoints ResourceExport and generate a new EndpointSubset
+	newSubsets := []corev1.EndpointSubset{}
+	undeleteItems, err := r.getUndeletedResourceExportItems(resExport)
+	if err != nil {
+		klog.ErrorS(err, "fail to list ResourceExports, retry later")
+		return newResImport, false, err
+	}
+	for _, re := range undeleteItems {
+		newSubsets = append(newSubsets, re.Spec.Endpoints.Subsets...)
+	}
+	newResImport.Spec.Endpoints = &mcsv1alpha1.EndpointsImport{Subsets: newSubsets}
+	if apiequality.Semantic.DeepEqual(newResImport.Spec.Endpoints, resImport.Spec.Endpoints) {
+		return newResImport, false, nil
+	}
+	return newResImport, true, nil
+}
+
+func (r *ResourceExportReconciler) getUndeletedResourceExportItems(resExport *mcsv1alpha1.ResourceExport) ([]mcsv1alpha1.ResourceExport, error) {
+	reList := &mcsv1alpha1.ResourceExportList{}
+	err := r.Client.List(context.TODO(), reList, &client.ListOptions{
+		LabelSelector: getLabelSelector(resExport),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return removeDeletedItems(reList.Items), nil
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *ResourceExportReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// ignore status update event via GenerationChangedPredicate
+	instance := predicate.GenerationChangedPredicate{}
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&multiclusterv1alpha1.ResourceExport{}).
+		For(&mcsv1alpha1.ResourceExport{}).
+		WithEventFilter(instance).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
 		Complete(r)
 }
+
+func getLabelSelector(resExport *mcsv1alpha1.ResourceExport) labels.Selector {
+	labelSelector := metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"sourceNamespace": resExport.Spec.Namespace,
+			"sourceName":      resExport.Spec.Name,
+			"sourceKind":      resExport.Spec.Kind,
+		},
+	}
+	selector, _ := metav1.LabelSelectorAsSelector(&labelSelector)
+	return selector
+}
+
+func svcPortsConverter(svcPort []corev1.ServicePort) []mcs.ServicePort {
+	var mcsSP []mcs.ServicePort
+	for _, v := range svcPort {
+		mcsSP = append(mcsSP, mcs.ServicePort{
+			Name:     strconv.Itoa(int(v.Port)) + strings.ToLower(string(v.Protocol)),
+			Port:     v.Port,
+			Protocol: v.Protocol,
+		})
+	}
+	return mcsSP
+}
+
+func getResourceImportName(resExport *mcsv1alpha1.ResourceExport) string {
+	return resExport.Spec.Namespace + "-" + resExport.Spec.Name + "-" + strings.ToLower(resExport.Spec.Kind)
```

Comment:
Why we need Namespace here?

Reply:
two Services from different Namespace might be exported by member clusters, they are actually different Services, so we need two different ResourceImport, otherwise, there will be name conflicts.